shell/packages/sandstorm-permissions/permissions.js

```diff
@@ -291,100 +252,173 @@ class RequirementSet {
       delete this.identityPermissions[grainId];
     }
 
-    return { vertex: { grain: { _id: grainId, identityId: identityId } },
-             permissions: permissions, };
+    return { grainId: grainId, identityId: identityId, permissionSet: permissions, };
+  }
+}
+
+const permissionIdPattern = Match.OneOf({ canAccess: null }, { appDefined: Match.Integer });
```

Comment:
This feels bulky to me. Possible alternative: Match.OneOf(Number, "canAccess"), i.e. the value is either a number or the special string value "canAccess". You can then represent a permission set as an object whose keys are numbers or "canAccess". This actually makes more sense than an array in that a permission array is really an infinite array with an all-false tail, so now we have an infinite object where unset fields are considered false...